src/libs/swaps/basis/rBasisSwap.ts

```diff
@@ -0,0 +1,613 @@
+/* eslint-disable new-cap */
+/* eslint-disable no-restricted-imports */
+import {
+  ASSOCIATED_TOKEN_PROGRAM_ID,
+  Token,
+  TOKEN_PROGRAM_ID,
+  AccountLayout,
+  u64,
+} from '@solana/spl-token';
+import {
+  Transaction,
+  Account,
+  Connection,
+  PublicKey,
+  TransactionInstruction,
+} from '@solana/web3.js';
+import BN from 'bn.js';
+import * as BufferLayout from 'buffer-layout';
+import { snakeCase } from 'snake-case';
+import { sha256 } from 'js-sha256';
+import * as Layout from '../../layout';
+import { StakingInstruction, StakingInstructionNames } from './instruction';
+
+// CONSTANTS
+// NOTE: did not use the usual getTokenInfo pattern as not sure how useful
+// an abstract redemption is as staking contracts aren't universal
+const SIGHASH_GLOBAL_NAMESPACE = 'global';
+export const MINT_BASIS = 'Basis9oJw9j8cw53oMV7iqsgo6ihi9ALw4QR31rcjUJa';
+export const MINT_RBASIS = 'rBsH9ME52axhqSjAVXY3t1xcCrmntVNvP3X16pRjVdM';
+export const PROGRAM_BASIS_STAKING = 'FTH1V7jAETZfDgHiL4hJudKXtV8tqKN1WEnkyY4kNAAC';
+export const PROGRAM_BASIS_STAKING_INSTANCE = 'HXCJ1tWowNNNUSrtoVnxT3y9ue1tkuaLNbFMM239zm1y';
+export const PROGRAM_BASIS_STAKING_VAULT = '3sBX8hj4URsiBCSRV26fEHkake295fQnM44EYKKsSs51';
+
+// NOTE: LIFTED FROM ANCHOR
+export function sighash(nameSpace: string, ixName: string): Buffer {
+  const name = snakeCase(ixName);
+  const preimage = `${nameSpace}:${name}`;
+  return Buffer.from(sha256.digest(preimage)).slice(0, 8);
+}
+
+function parseTokenAccountData(account, data) {
+  const accountInfo = AccountLayout.decode(data);
+  accountInfo.address = account;
+  accountInfo.mint = new PublicKey(accountInfo.mint);
+  accountInfo.owner = new PublicKey(accountInfo.owner);
+  accountInfo.amount = u64.fromBuffer(accountInfo.amount);
+
+  if (accountInfo.delegateOption === 0) {
+    accountInfo.delegate = null;
+    accountInfo.delegatedAmount = new u64(0);
+  } else {
+    accountInfo.delegate = new PublicKey(accountInfo.delegate);
+    accountInfo.delegatedAmount = u64.fromBuffer(accountInfo.delegatedAmount);
+  }
+
+  accountInfo.isInitialized = accountInfo.state !== 0;
+  accountInfo.isFrozen = accountInfo.state === 2;
+
+  if (accountInfo.isNativeOption === 1) {
+    accountInfo.rentExemptReserve = u64.fromBuffer(accountInfo.isNative);
+    accountInfo.isNative = true;
+  } else {
+    accountInfo.rentExemptReserve = null;
+    accountInfo.isNative = false;
+  }
+
+  if (accountInfo.closeAuthorityOption === 0) {
+    accountInfo.closeAuthority = null;
+  } else {
+    accountInfo.closeAuthority = new PublicKey(accountInfo.closeAuthority);
+  }
+
+  return accountInfo;
+}
+
+export async function getTokenAccount(connection, publicKey) {
+  const result = await connection.getAccountInfo(publicKey);
+  if (!result) return false;
+  const data = Buffer.from(result.data);
+  const account = parseTokenAccountData(publicKey, data);
+  return {
+    publicKey,
+    account,
+  };
+}
```

Comment:
can we move parseTokenAccountData and getTokenAccount to src/libs/utils.ts. Seem like handy methods